File: ServiceExercise_1/app/src/main/java/com/programming/android/sdu/serviceexercise/JokeAndroidService.java
package com.programming.android.sdu.serviceexercise;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;
import androidx.annotation.Nullable;
import android.text.Html;
import android.util.Log;

import java.io.IOException;

import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by grzegorzbaczek on 18/03/2018.
 */

public class JokeAndroidService extends Service {

    private static final String url = "http://api.icndb.com/";

    private Thread workerThread;
    private JokeWebService jokeWebService;
    private volatile boolean running;
    private long jokeCounter;
    private String newestJoke;
    // Binder given to clients
    private final IBinder binder = new JokeAndroidServiceBinder();


    @Override
    public void onCreate() {
        super.onCreate();
        Log.i("service_exrecise", "JokeAndroidService onCreate- Current Thread ID- " + Thread.currentThread().getId() + " For Thread- " + Thread.currentThread().getName());
        jokeCounter = 0;
        running = true;
        Retrofit retrofit = new Retrofit.Builder().baseUrl(url)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        jokeWebService = retrofit.create(JokeWebService.class);
        workerThread = new Thread(new Runnable() {
            @Override
            public void run() {
                while (running) {
                    Call<Joke> joke = jokeWebService.randomJoke();
                    try {
                        String joke_txt = Html.fromHtml(joke.execute().body().getValue().getJoke()).toString();
                        newestJoke = joke_txt;
                        jokeCounter++;
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                    try {
                        Thread.sleep(3000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        });

        workerThread.start();

    }

    public String getNewestJoke(){
        return newestJoke;
    }

    public Long getJokeCounter(){
        return jokeCounter;
    }

    public IBinder onBind(Intent intent) {
        Log.i("service_exrecise", "JokeAndroidService onBind- Current Thread ID- " + Thread.currentThread().getId() + " For Thread- " + Thread.currentThread().getName());
        return binder;
    }

    public class JokeAndroidServiceBinder extends Binder {
        JokeAndroidService getService() {
            // Return this instance of LocalService so clients can call public methods
            return JokeAndroidService.this;
        }
    }

    @Override
    public void onDestroy() {
        running = false;
        try {
            workerThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        Log.i("service_exrecise", "JokeAndroidService onDestroy- Current Thread ID- " + Thread.currentThread().getId() + " For Thread- " + Thread.currentThread().getName());
        super.onDestroy();

    }
}

File: Android 10 - Android services, Work managers - executing long running tasks/ServiceExercise_1/app/src/main/java/com/programming/android/sdu/serviceexercise/JokeAndroidService.kt
package com.programming.android.sdu.serviceexercise

import android.app.Service
import android.content.Intent
import android.os.Binder
import android.os.IBinder
import android.text.Html
import android.util.Log
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException

/**
 * Created by grzegorzbaczek on 18/03/2018.
 */
class JokeAndroidService : Service() {

    private var workerThread: Thread? = null
    private val url = "https://api.icndb.com/"

    @Volatile
    private var running = false
    var jokeCounter: Long = 0
        private set
    var newestJoke: String? = null
        private set

    // Binder given to clients
    private val binder: IBinder = JokeAndroidServiceBinder()

    override fun onCreate() {
        super.onCreate()
        Log.i("service_exrecise", "JokeAndroidService onCreate- Current Thread ID- " + Thread.currentThread().id + " For Thread- " + Thread.currentThread().name)
        jokeCounter = 0
        running = true
        val retrofit = Retrofit.Builder().baseUrl(url)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        val jokeWebService = retrofit.create(JokeWebService::class.java)
        workerThread = Thread {
            while (running) {
                val joke = jokeWebService.randomJoke()
                try {
                    val joke_txt = Html.fromHtml(joke!!.execute().body()!!.value!!.joke).toString()
                    newestJoke = joke_txt
                    jokeCounter++
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                try {
                    Thread.sleep(3000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }
        workerThread!!.start()
    }

    override fun onBind(intent: Intent): IBinder {
        Log.i("service_exrecise", "JokeAndroidService onBind- Current Thread ID- " + Thread.currentThread().id + " For Thread- " + Thread.currentThread().name)
        return binder
    }

    inner class JokeAndroidServiceBinder : Binder() {
        // Return this instance of LocalService so clients can call public methods
        val service: JokeAndroidService
            get() =// Return this instance of LocalService so clients can call public methods
                this@JokeAndroidService
    }

    override fun onDestroy() {
        running = false
        try {
            workerThread!!.join()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        Log.i("service_exrecise", "JokeAndroidService onDestroy- Current Thread ID- " + Thread.currentThread().id + " For Thread- " + Thread.currentThread().name)
        super.onDestroy()
    }

}